designs/2019-worker-threads/README.md

```diff
@@ -0,0 +1,338 @@
+
+#### About `--fix` option
+
+We can use parallel linting along with `--fix` option. ESLint modifies files after all linting completed.
+
+#### About `--debug` option
+
+We can use parallel linting along with `--debug` option, but ESLint cannot guarantee the order of the debug log. The order of the logs that are received from worker threads is messed.
+
+#### About Error Handling
+
+If exceptions happen in worker threads, the worker transfers the error information to the main thread if possible. The transferred information contains the properties of `message`, `stack`, `messageTemplate`, and `messageData`.
+
+Therefore, people can see the error message as-is even if it happened in worker threads.
+For example:
+
+```
+$ eslint lib
+Error: FAILED!!!
+Occurred while linting C:\Users\t-nagashima.AD\dev\eslint\lib\rules\no-mixed-operators.js:229
+    at BinaryExpression (C:\Users\t-nagashima.AD\dev\eslint\lib\rules\eqeqeq.js:152:27)
+    at C:\Users\t-nagashima.AD\dev\eslint\lib\linter\safe-emitter.js:45:58
+    at Array.forEach (<anonymous>)
+    at Object.emit (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\safe-emitter.js:45:38)
+    at NodeEventGenerator.applySelector (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\node-event-generator.js:253:26)
+    at NodeEventGenerator.applySelectors (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\node-event-generator.js:282:22)
+    at NodeEventGenerator.enterNode (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\node-event-generator.js:296:14)
+    at CodePathAnalyzer.enterNode (C:\Users\t-nagashima.AD\dev\eslint\lib\linter\code-path-analysis\code-path-analyzer.js:646:23)
+    at C:\Users\t-nagashima.AD\dev\eslint\lib\linter\linter.js:936:32
+    at Array.forEach (<anonymous>)
+```
+
+As a side note, master terminates the other workers if a worker reported an error.
+
+#### About aborting
+
+In [RFC 40](https://github.com/eslint/rfcs/pull/40), `ESLint#executeOnFiles()` returns async iterable object.
+
+The iterator has optional [`return()` method](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator/return). The `for-of`/`for-await-of` syntax calls the `return()` method automatically if the execution escaped from the loop by `braek`, `return`, or `throw`. In short, the `return()` method will be called when aborted.
+
+Therefore, ESLint aborts linting when the `return()` method is called. This means that the `return()` method terminates all workers.
+
+```js
+const { ESLint } = require("eslint")
+const eslint = new ESLint()
+for await (const result of eslint.executeOnFiles(patterns)) {
+  if (Math.random() < 0.5) {
+    break // abort linting -- terminate all workers if running.
+  }
+}
+```
+
+And it throws an error if it's reused after aborted.
```

Comment:
Maybe just replace this with an explanation that the async iterables are possible to abort and that any aborting will result in all the workers getting shut down?

The inner workings of what, how and why is better done in the referenced RFC I think 🙂

Reply:
Because I thought people image [the `AbortController` standard](https://dom.spec.whatwg.org/#aborting-ongoing-activities) at first to abort asynchronous operations. The fact that async iteration has another mechanism is grounds for we don't use `AbortController`.